File: android/src/main/java/expo/modules/mediaedit/ExpoMediaEditModule.kt
package expo.modules.mediaedit

import android.media.MediaExtractor
import android.media.MediaFormat
import android.media.MediaMetadataRetriever
import android.net.Uri
import android.os.Bundle
import androidx.media3.common.util.UnstableApi
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.File

@UnstableApi
class ExpoMediaEditModule : Module() {

  @Volatile private var activeExporter: ProjectExporter? = null

  override fun definition() = ModuleDefinition {
    Name("ExpoMediaEdit")

    Events("onProgress")

    // MARK: - exportProject(project, outputUri?, opts?)
    AsyncFunction("exportProject") { projectMap: Map<String, Any?>, outputUri: String?, opts: Map<String, Any?>?, promise: Promise ->
      val ctx = requireNotNull(appContext.reactContext) { "React context not available" }

      val outputFile = if (outputUri != null && outputUri.isNotEmpty()) {
        resolveAllowedOutputFile(ctx, outputUri)
          ?: run {
            promise.reject("INVALID_OUTPUT", "outputUri must be a file:// path inside the app's sandbox", null)
            return@AsyncFunction
          }
      } else {
        createTempFile(ctx, "output", ".mp4")
      }

      val quality = (opts?.get("quality") as? String) ?: "high"

      Thread {
        val compiled = try {
          val project = ProjectParser.parse(projectMap)
          ProjectCompiler.compile(ctx, project, ProjectCompiler.Mode.EXPORT)
        } catch (e: Exception) {
          promise.reject("COMPILE_FAILED", "${e.javaClass.simpleName}: ${e.message ?: ""}", e)
          return@Thread
        }

        val exporter = ProjectExporter(ctx)
        activeExporter = exporter
        exporter.export(
          composition = compiled.composition,
          outputFile = outputFile,
          quality = quality,
          onProgress = { p ->
            sendEvent("onProgress", Bundle().apply { putDouble("progress", p.toDouble()) })
          },
          completion = { result ->
            activeExporter = null
            result.fold(
              onSuccess = { f -> promise.resolve("file://${f.absolutePath}") },
              onFailure = { e ->
                if (e is CancellationException) promise.reject("CANCELLED", e.message ?: "cancelled", e)
                else promise.reject("EXPORT_FAILED", e.message ?: "Unknown export error", e)
              }
            )
          }
        )
      }.start()
    }

    AsyncFunction("cancelExport") { promise: Promise ->
      activeExporter?.cancel()
      promise.resolve(null)
    }

    // MARK: - getVideoInfo / generateThumbnail / extractAudio / cleanTempFiles
    // Unchanged from 0.13.x — independent of the composer.

    AsyncFunction("getVideoInfo") { uri: String, promise: Promise ->
      if (!isReadableUriAllowed(uri)) {
        promise.reject("INVALID_URI", "uri must be a file:// or https:// URI without path traversal", null)
        return@AsyncFunction
      }
      val ctx = requireNotNull(appContext.reactContext)
      val retriever = MediaMetadataRetriever()
      try {
        retriever.setDataSource(ctx, Uri.parse(uri))
        val durationMs = retriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_DURATION)?.toLong() ?: 0L
        val rawWidth = retriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_VIDEO_WIDTH)?.toInt() ?: 0
        val rawHeight = retriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_VIDEO_HEIGHT)?.toInt() ?: 0
        val rotation = retriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_VIDEO_ROTATION)?.toInt() ?: 0
        val (width, height) = if (rotation == 90 || rotation == 270) rawHeight to rawWidth else rawWidth to rawHeight
        val fileSize = if (uri.startsWith("file://")) File(uri.removePrefix("file://")).length() else 0L

        var fps = 0f
        var codecMime: String? = null
        val extractor = MediaExtractor()
        try {
          extractor.setDataSource(ctx, Uri.parse(uri), null)
          for (i in 0 until extractor.trackCount) {
            val format = extractor.getTrackFormat(i)
            val mime = format.getString(MediaFormat.KEY_MIME) ?: continue
            if (mime.startsWith("video/")) {
              codecMime = mime
              if (format.containsKey(MediaFormat.KEY_FRAME_RATE)) {
                fps = try { format.getInteger(MediaFormat.KEY_FRAME_RATE).toFloat() }
                      catch (_: Exception) { format.getFloat(MediaFormat.KEY_FRAME_RATE) }
              }
              break
            }
          }
        } catch (_: Exception) {
          fps = retriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_CAPTURE_FRAMERATE)?.toFloat() ?: 0f
        } finally {
          extractor.release()
        }
        val codec = mimeToFourCC(codecMime)

        promise.resolve(mapOf(
          "durationMs" to durationMs.toDouble(),
          "width" to width.toDouble(),
          "height" to height.toDouble(),
          "fps" to fps.toDouble(),
          "fileSize" to fileSize.toDouble(),
          "codec" to codec
        ))
      } finally {
        retriever.release()
      }
    }

    AsyncFunction("generateThumbnail") { uri: String, timeMs: Double, options: Map<String, Any?>?, promise: Promise ->
      if (!isReadableUriAllowed(uri)) {
        promise.reject("INVALID_URI", "uri must be a file:// or https:// URI without path traversal", null)
        return@AsyncFunction
      }
      val ctx = requireNotNull(appContext.reactContext)
      val retriever = MediaMetadataRetriever()
      try {
        retriever.setDataSource(ctx, Uri.parse(uri))
        val bitmap = retriever.getFrameAtTime(
          (timeMs * 1000).toLong(),
          MediaMetadataRetriever.OPTION_CLOSEST_SYNC
        )
        if (bitmap != null) {
          val outputFile = createTempFile(ctx, "thumb", ".jpg")
          outputFile.outputStream().use { out ->
            bitmap.compress(android.graphics.Bitmap.CompressFormat.JPEG, 80, out)
          }
          bitmap.recycle()
          promise.resolve("file://${outputFile.absolutePath}")
        } else {
          promise.reject("THUMBNAIL_FAILED", "Could not extract frame at ${timeMs}ms", null)
        }
      } finally {
        retriever.release()
      }
    }

    AsyncFunction("extractAudio") { uri: String, promise: Promise ->
      if (!isReadableUriAllowed(uri)) {
        promise.reject("INVALID_URI", "uri must be a file:// or https:// URI without path traversal", null)
        return@AsyncFunction
      }
      val ctx = requireNotNull(appContext.reactContext)
      Thread {
        try {
          val extractor = MediaExtractor()
          extractor.setDataSource(ctx, Uri.parse(uri), null)
          var audioTrack = -1
          var audioFormat: MediaFormat? = null
          for (i in 0 until extractor.trackCount) {
            val format = extractor.getTrackFormat(i)
            val mime = format.getString(MediaFormat.KEY_MIME) ?: continue
            if (mime.startsWith("audio/")) {
              audioTrack = i; audioFormat = format; break
            }
          }
          if (audioTrack < 0 || audioFormat == null) {
            extractor.release()
            promise.reject("NO_AUDIO_TRACK", "Source has no audio track", null)
            return@Thread
          }
          extractor.selectTrack(audioTrack)
          val outputFile = createTempFile(ctx, "audio", ".m4a")
          val muxer = android.media.MediaMuxer(outputFile.absolutePath, android.media.MediaMuxer.OutputFormat.MUXER_OUTPUT_MPEG_4)
          val muxAudioTrack = muxer.addTrack(audioFormat)
          muxer.start()
          val bufferSize = audioFormat.getInteger(MediaFormat.KEY_MAX_INPUT_SIZE)
            .let { if (it > 0) it else 1024 * 1024 }
          val buffer = java.nio.ByteBuffer.allocate(bufferSize)
          val bufferInfo = android.media.MediaCodec.BufferInfo()
          while (true) {
            val sampleSize = extractor.readSampleData(buffer, 0)
            if (sampleSize < 0) break
            bufferInfo.offset = 0
            bufferInfo.size = sampleSize
            bufferInfo.presentationTimeUs = extractor.sampleTime
            bufferInfo.flags = extractor.sampleFlags
            muxer.writeSampleData(muxAudioTrack, buffer, bufferInfo)
            extractor.advance()
          }
          muxer.stop(); muxer.release(); extractor.release()
          promise.resolve("file://${outputFile.absolutePath}")
        } catch (e: Exception) {
          promise.reject("EXTRACT_FAILED", e.message ?: "Audio extraction failed", e)
        }
      }.start()
    }

    AsyncFunction("cleanTempFiles") { promise: Promise ->
      val ctx = requireNotNull(appContext.reactContext)
      val tempDir = File(ctx.cacheDir, "expo-media-edit")
      var count = 0
      tempDir.listFiles()?.forEach { file -> if (file.delete()) count++ }
      promise.resolve(count)
    }

    // MARK: - <MediaPreview /> view

    View(MediaPreviewView::class) {
      Events("onTime", "onReady", "onError")

      Prop("project") { view: MediaPreviewView, project: Map<String, Any?> ->
        view.updateProject(project)
      }
      Prop("time") { view: MediaPreviewView, time: Double ->
        view.updateTime(time)
      }
      Prop("playing") { view: MediaPreviewView, playing: Boolean ->
        view.updatePlaying(playing)
      }
      Prop("renderScale") { view: MediaPreviewView, scale: Double ->
        view.updateRenderScale(scale)
      }
    }
  }
}

class CancellationException(message: String) : Exception(message)

private fun mimeToFourCC(mime: String?): String? = when (mime) {
  null -> null
  "video/avc" -> "avc1"
  "video/hevc" -> "hvc1"
  "video/x-vnd.on2.vp8" -> "vp08"
  "video/x-vnd.on2.vp9" -> "vp09"
  "video/av01" -> "av01"
  "video/mp4v-es" -> "mp4v"
  else -> mime
}

File: ios/ExpoMediaEditModule.swift
import ExpoModulesCore
import AVFoundation

public class ExpoMediaEditModule: Module {
  private var activeExporter: ProjectExporter?

  public func definition() -> ModuleDefinition {
    Name("ExpoMediaEdit")

    Events("onProgress")

    // MARK: - exportProject(project, outputUri?, opts?)
    //
    // Replaces the 0.13.x editVideo(job). Takes a Project dict (same
    // shape as src/project.ts), compiles it via ProjectCompiler, and
    // runs the result through ProjectExporter to an MP4 on disk.
    AsyncFunction("exportProject") { (projectDict: [String: Any], outputUri: String?, opts: [String: Any]?, promise: Promise) in
      let outputURL: URL
      if let uri = outputUri, !uri.isEmpty {
        guard let url = URL(string: uri), MediaEditSecurity.isOutputURLAllowed(url) else {
          promise.reject("INVALID_OUTPUT", "outputUri must be a file:// path inside the app's sandbox")
          return
        }
        outputURL = url
      } else {
        let tempDir = FileManager.default.temporaryDirectory.appendingPathComponent("expo-media-edit")
        try? FileManager.default.createDirectory(at: tempDir, withIntermediateDirectories: true)
        outputURL = tempDir.appendingPathComponent(UUID().uuidString + ".mp4")
      }

      let quality = (opts?["quality"] as? String) ?? "high"

      DispatchQueue.global(qos: .userInitiated).async { [weak self] in
        guard let self = self else { return }
        let compiled: CompiledComposition
        do {
          let project = try ProjectParser.parse(projectDict)
          compiled = try ProjectCompiler.compile(project, mode: .export)
        } catch {
          promise.reject("COMPILE_FAILED", "\(error)")
          return
        }

        let exporter = ProjectExporter()
        self.activeExporter = exporter
        exporter.export(
          compiled: compiled,
          outputURL: outputURL,
          quality: quality,
          onProgress: { [weak self] p in
            self?.sendEvent("onProgress", ["progress": p])
          },
          completion: { [weak self] result in
            self?.activeExporter = nil
            switch result {
            case .success(let url):
              promise.resolve(url.absoluteString)
            case .failure(let err):
              let nsErr = err as NSError
              let msg = nsErr.localizedDescription
              if msg.contains("cancelled") || msg.contains("Cancelled") {
                promise.reject("CANCELLED", msg)
              } else {
                promise.reject("EXPORT_FAILED", msg)
              }
            }
          }
        )
      }
    }

    AsyncFunction("cancelExport") { (promise: Promise) in
      self.activeExporter?.cancel()
      promise.resolve(nil)
    }

    // MARK: - getVideoInfo / generateThumbnail / extractAudio / cleanTempFiles
    //
    // Unchanged from 0.13.x — these are independent of the composer
    // and operate on individual files.

    AsyncFunction("getVideoInfo") { (uri: String, promise: Promise) in
      guard MediaEditSecurity.isReadableURIAllowed(uri) else {
        promise.reject("INVALID_URI", "uri must be a file:// or https:// URI without path traversal"); return
      }
      guard let url = URL(string: uri) else {
        promise.reject("INVALID_URI", "uri is not a valid URL"); return
      }
      let asset = AVAsset(url: url)
      let track = asset.tracks(withMediaType: .video).first
      let fileSize = (try? url.resourceValues(forKeys: [.fileSizeKey]).fileSize) ?? 0

      var width = Double(track?.naturalSize.width ?? 0)
      var height = Double(track?.naturalSize.height ?? 0)
      if let t = track {
        let rendered = t.naturalSize.applying(t.preferredTransform)
        width = Double(abs(rendered.width))
        height = Double(abs(rendered.height))
      }

      var codec: String? = nil
      if let formatDescs = track?.formatDescriptions as? [CMFormatDescription],
         let first = formatDescs.first {
        let fourCC = CMFormatDescriptionGetMediaSubType(first)
        codec = String(format: "%c%c%c%c",
                       (fourCC >> 24) & 0xff,
                       (fourCC >> 16) & 0xff,
                       (fourCC >> 8) & 0xff,
                       fourCC & 0xff)
      }

      var result: [String: Any] = [
        "durationMs": CMTimeGetSeconds(asset.duration) * 1000.0,
        "width": width, "height": height,
        "fps": Double(track?.nominalFrameRate ?? 0),
        "fileSize": fileSize,
      ]
      if let c = codec { result["codec"] = c }
      promise.resolve(result)
    }

    AsyncFunction("generateThumbnail") { (uri: String, timeMs: Double, options: [String: Any]?, promise: Promise) in
      guard MediaEditSecurity.isReadableURIAllowed(uri) else {
        promise.reject("INVALID_URI", "uri must be a file:// or https:// URI without path traversal"); return
      }
      guard let url = URL(string: uri) else {
        promise.reject("INVALID_URI", "uri is not a valid URL"); return
      }
      let asset = AVAsset(url: url)
      let generator = AVAssetImageGenerator(asset: asset)
      generator.appliesPreferredTrackTransform = true
      if let w = options?["width"] as? Double, let h = options?["height"] as? Double {
        generator.maximumSize = CGSize(width: w, height: h)
      }
      let time = CMTime(value: CMTimeValue(timeMs), timescale: 1000)
      do {
        let cgImage = try generator.copyCGImage(at: time, actualTime: nil)
        let image = UIImage(cgImage: cgImage)
        let tempDir = FileManager.default.temporaryDirectory.appendingPathComponent("expo-media-edit")
        try FileManager.default.createDirectory(at: tempDir, withIntermediateDirectories: true)
        let outputURL = tempDir.appendingPathComponent(UUID().uuidString + ".jpg")
        if let data = image.jpegData(compressionQuality: 0.8) {
          try data.write(to: outputURL)
          promise.resolve(outputURL.absoluteString)
        } else {
          promise.reject("THUMBNAIL_FAILED", "Could not encode thumbnail as JPEG")
        }
      } catch {
        promise.reject("THUMBNAIL_FAILED", error.localizedDescription)
      }
    }

    AsyncFunction("extractAudio") { (uri: String, promise: Promise) in
      guard MediaEditSecurity.isReadableURIAllowed(uri) else {
        promise.reject("INVALID_URI", "uri must be a file:// or https:// URI without path traversal"); return
      }
      guard let url = URL(string: uri) else {
        promise.reject("INVALID_URI", "uri is not a valid URL"); return
      }
      let asset = AVAsset(url: url)
      guard !asset.tracks(withMediaType: .audio).isEmpty else {
        promise.reject("NO_AUDIO_TRACK", "Source has no audio track"); return
      }
      guard let exporter = AVAssetExportSession(asset: asset, presetName: AVAssetExportPresetAppleM4A) else {
        promise.reject("EXPORT_INIT_FAILED", "Could not create AVAssetExportSession"); return
      }
      let tempDir = FileManager.default.temporaryDirectory.appendingPathComponent("expo-media-edit")
      try? FileManager.default.createDirectory(at: tempDir, withIntermediateDirectories: true)
      let outputURL = tempDir.appendingPathComponent(UUID().uuidString + ".m4a")
      exporter.outputURL = outputURL
      exporter.outputFileType = .m4a
      exporter.exportAsynchronously {
        switch exporter.status {
        case .completed: promise.resolve(outputURL.absoluteString)
        case .failed, .cancelled:
          promise.reject("EXTRACT_FAILED", exporter.error?.localizedDescription ?? "Audio extraction failed")
        default:
          promise.reject("EXTRACT_FAILED", "Unexpected exporter status: \(exporter.status.rawValue)")
        }
      }
    }

    AsyncFunction("cleanTempFiles") { (promise: Promise) in
      let tempDir = FileManager.default.temporaryDirectory.appendingPathComponent("expo-media-edit")
      let files = try? FileManager.default.contentsOfDirectory(at: tempDir, includingPropertiesForKeys: nil)
      var count = 0
      files?.forEach { url in
        if (try? FileManager.default.removeItem(at: url)) != nil { count += 1 }
      }
      promise.resolve(count)
    }

    // MARK: - <MediaPreview /> view
    //
    // Native view component. Receives `project` JSON, `time`,
    // `playing`, optional `renderScale`. Emits `onTime`, `onReady`,
    // `onError`. Internally builds an AVPlayer fed by the same
    // ProjectCompiler output that the exporter uses → preview pixels
    // == export pixels.
    View(MediaPreviewView.self) {
      Events("onTime", "onReady", "onError")

      Prop("project") { (view: MediaPreviewView, project: [String: Any]) in
        view.updateProject(project)
      }
      Prop("time") { (view: MediaPreviewView, time: Double) in
        view.updateTime(time)
      }
      Prop("playing") { (view: MediaPreviewView, playing: Bool) in
        view.updatePlaying(playing)
      }
      Prop("renderScale") { (view: MediaPreviewView, scale: Double) in
        view.updateRenderScale(scale)
      }
    }
  }
}
